### PR TITLE
Update NPM packaging, linting, and a bunch of types

### PR DESCRIPTION
**Pull request by emma-sg**
Preparation for https://github.com/webrecorder/archiveweb.page/issues/198 / https://github.com/webrecorder/archiveweb.page/pull/199.

## Changes

- Organizes published files more cleanly, using `dist/types` folder as the target for type declarations and sourcemaps
- Configures type field in `package.json`
- Updates ESLint & its config
  - Sets up type-informed linting rules mirroring browsertrix
  - Adds a bunch of other useful eslint and eslint-typescript rules, and applies auto-fixes to the codebase
- Cleans up a bunch of types throughout the codebase, particularly parts of the code used by ArchiveWebPage
  - Changes `ItemIndex#_deleting` from `private` to `protected` so that a part of AWP that extends it can do so properly
  - Makes a bunch of previously-implied [non-null assertions](https://www.typescriptlang.org/docs/handbook/2/everyday-types.html#non-null-assertion-operator-postfix-) explicit
  - Updates `URLResource` and `ItemType` types (and renames `Item` → `ItemType` to disambiguate from the `Item` class)
- Removes `node-sass` and replaces it with `sass-embedded` for much quicker installs and compilation

## Changes to review that may impact functionality

- Type `swErrorMsg` as a `String` rather than an `Object`
  - https://github.com/webrecorder/replayweb.page/pull/275/files#diff-95446563bbca1be542e6c6bf249f786764134fe8052a46cfaf0ec98ec28fb512R73

- Explicitly convert a couple things to `boolean`s, rather than relying on implicit conversions that could be `undefined`s or `null`s
  - https://github.com/webrecorder/replayweb.page/pull/275/files#diff-8f6033fd1f4a1e06eddb4e0a51ad9f49e7d2543fe754c726263c1d04f2e6aab0R91
  - https://github.com/webrecorder/replayweb.page/pull/275/files#diff-0a1342822ff2e711a79fd7036fdc352146751bd42d7d5845a2938f903723f74bR273
 
- Change a couple of `||`s to `??`s
  - https://github.com/webrecorder/replayweb.page/pull/275/files#diff-839757414c902c8de340ce583258acff5298dd16e1c88935c18f8c6a7df3396aR186
  - https://github.com/webrecorder/replayweb.page/pull/275/files#diff-8f6033fd1f4a1e06eddb4e0a51ad9f49e7d2543fe754c726263c1d04f2e6aab0R107

- Remove `async` from a function that doesn't need it
  - https://github.com/webrecorder/replayweb.page/pull/275/files#diff-0a1342822ff2e711a79fd7036fdc352146751bd42d7d5845a2938f903723f74bR242
  - There are a few more cases of this remaining in the codebase, but I wasn't totally sure if removing `async` from them might change an aspect of the outward API shape
 
- Use the `state` property of the `PermissionStatus` object returned by `FileSystemHandle#requestPermission`, rather than using it as a string. I'm not sure if this was properly working before or not?
  - https://github.com/webrecorder/replayweb.page/pull/275/files#diff-0a1342822ff2e711a79fd7036fdc352146751bd42d7d5845a2938f903723f74bR426
  - https://github.com/webrecorder/replayweb.page/pull/275/files#diff-02dbd14e32846aa161148c62042c9d8f19db0e42fec5a26e189fc9b1b908730dR66

## Testing
I've been poking around locally and haven't found issues, but I'd appreciate some help here!

### CODE REVIEW

File: .eslintrc.js [hunk 1]
```diff
@@ -1,32 +1,110 @@
...
-  parser: "@typescript-eslint/parser",
   parserOptions: {
-    ecmaVersion: 2018,
-    sourceType: "module",
+    project: ["./tsconfig.eslint.json"],
+    tsconfigRootDir: __dirname,
   },
+  root: true,
   rules: {
+    /* start stylistic rules */
```
Comment (SuaYoo) [inline .eslintrc.js:24]: Are the additional rules here mostly migration purposes?

File: src/embed-receipt.ts [hunk 2]
```diff
@@ -183,15 +183,15 @@ export class RWPEmbedReceipt extends LitElement {
...
+    } = this.collInfo?.verify ?? {};
 
     const sourceUrl = this.collInfo?.sourceUrl;
 
     const certFingerprintUrl = certFingerprint
       ? `https://crt.sh/?q=${certFingerprint}`
       : "";
 
-    const dateStr = tsToDate(this.ts).toLocaleString();
+    const dateStr = tsToDate(this.ts!).toLocaleString();
```
Comment (SuaYoo) [inline src/embed-receipt.ts:194]: Will this throw if `this.ts` is missing for whatever reason? Maybe we can handle this in the template, `if (!this.ts) return nothing`

File: .eslintrc.js (hunk 1 above)
Comment (emma-sg) [reply .eslintrc.js:24]: Yeah, that and keeping code quality high moving forward as well — most of these rules have auto-fixable suggestions, so I think this should be a good improvement in that regard without much added friction

File: src/embed-receipt.ts (hunk 2 above)
Comment (emma-sg) [reply src/embed-receipt.ts:194]: Good question! It actually looks like `tsToDate` already returns an empty string if passed a falsy value, lemme just update that type.

**Comment by ikreymer** [comment]
> Use the state property of the PermissionStatus object returned by FileSystemHandle#requestPermission, rather than using it as a string. I'm not sure if this was properly working before or not?

It appears that Chrome does actually return a string here, and that seems to be what the current spec draft calls for:
https://wicg.github.io/file-system-access/#api-filesystemhandle-requestpermission

However, MDN lists that it returns PermissionStatus, so there is some discrepancy. For now, need to change it back to string to make it work.

**Review by ikreymer — APPROVED** [review]

**Pull request merged**